Emoncms feeds sensor component

joyrider3774 opened the pull request:
**Description:**
Emoncms is a powerful open-source web-app for processing, logging and visualising energy, temperature and other environmental data.  More info here https://emoncms.org/

This component can create sensors for all or specific feeds on the platform using their could based install or a local install. One get all feeds, specific feeds specified by feedid, or all feeds except the feeds mentioned in the exception list, as a sensor. I have tested it against my local install, cloud based install. Another home assistant forum user tested it against the lasted version of emon cms in a docker version

default naming sheme is "emoncms[sensorid]_feedid_[emoncmsfeedid]" for the senor but can be changed as mentioned earlier when using "include_feed_id" with "include_feed_id_names". if less elements are given  for "include_feed_id_names" the other sensors will use default naming scheme
The id config value is used in the default namening scheme ([sensorid] above) to better identify the sensors and be sure name will be unique.

Can be used against multiple emoncms instance but needs a config entry for each of them

Url when using the could based version should be https://emoncms.org

**Related issue (if applicable):** fixes #
none

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#
https://github.com/home-assistant/home-assistant.github.io/pull/910

**Example entry for `configuration.yaml` (if applicable):**

all feeds without unit_of_measurement specified (will be defaulted to "W") 

``` yaml
sensor:
    - platform: emoncms
      api_key: put your emoncms read api key here
      url: http://192.168.0.101:8083/emoncmsnew
      id: 1
```

single group of feeds by specifing feed id's using  include_only_feed_id

``` yaml
sensor:
    - platform: emoncms
      api_key: put your emoncms read api key here
      url: https://emoncms.org      
      id: 1
      unit_of_measurement: "W" 
      include_only_feed_id:
        - 107
        - 106
        - 105
```

all feed ids except feeds specified in exclude_feed_id

``` yaml
sensor:
    - platform: emoncms
      api_key: put your emoncms read api key here
      url: http://192.168.0.101:8083/emoncmsnew
      id: 1
      unit_of_measurement: "KWH" 
      exclude_feed_id:
        - 107
        - 106
        - 105
```

specify names for some of the sensors for the feeds specified in "include_only_feed_id" using "sensor_names", the other one will use my default naming scheme

``` yaml
sensor:
  - platform: emoncms
    api_key: put your emoncms read api key here
    url: http://192.168.0.101:8083/emoncmsnew
    id: 1
    unit_of_measurement: "W" 
    include_only_feed_id:
      - 5
      - 18
      - 29
      - 48
      - 61
      - 110
      - 116
      - 120
    sensor_names:
      5: "feed 1"
      18: "feed 2"
      29: "Totall watt"
```

use a "value_template" to multiply the feed value times 5 for all specified feed id's in "include_only_feed_id"

``` yaml
sensor:
  - platform: emoncms
    api_key: put your emoncms read api key here
    url: http://192.168.0.101:8083/emoncmsnew
    scan_interval: 15
    id: 1
    unit_of_measurement: "W"
    value_template: "{{ value | float * 5 }}"
    include_only_feed_id:
      - 107
      - 106
      - 105
      - 61
```

Same emoncms instance with 2 groups of feeds, diffrent scan_interval and a diffrent unit_of_measurement 

``` yaml
sensor:
    - platform: emoncms
      api_key: put your emoncms read api key here
      url: http://192.168.0.101:8083/emoncmsnew
      scan_interval: 5
      id: 1
      unit_of_measurement: "W" 
      include_only_feed_id:
        - 107
        - 106
        - 105
    - platform: emoncms
      api_key:  put your emoncms read api key here
      url: http://192.168.0.101:8083/emoncmsnew
      id: 2
      scan_interval: 60
      unit_of_measurement: "KW" 
      include_only_feed_id:
        - 108
        - 109
        - 110
        - 61
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices, web services, or a:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
- [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
- [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [X] New files were added to `.coveragerc`.

joyrider3774 committed:
Add Initial version for emoncms feeds sensor

joyrider3774 committed:
flake8 test fixes

joyrider3774 committed:
pylint test fixes

joyrider3774 committed:
- fix bug with include_feed_id_names not assigning the name to the  element in the same postion as found in include_feed_id
- a few structure changes to have less nesting (pylint fix)
- minor other changes

joyrider3774 committed:
update .coveragerc

kellerza commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,253 @@
+    exclude_length = len(exclude_feeds)
+    include_names_length = len(include_feeds_names)
+
+    if include_length > 0 and exclude_length > 0:
```
You can check this with voluptuous as well: 
```python
a=vol.Schema({
    vol.Exclusive('include', 'in_out_group'): str, 
    vol.Exclusive('exclude', 'in_out_group'): str
})
```

and possibly `cv.has_at_least_one_key` if you need at least one

joyrider3774 committed:
voluptuous fixes:
- exclude_feed_id and include_feed_id Exclusive group so that only one (or none) can be specified at once
- id must be positive int
- exclude_feed_id and include_feed_id must be positive int

kellerza commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+                              "'%s' is correct! (HTTP Status_code = %d)",
+                              CONF_URL, req.status_code)
+        finally:
+            if self.data is not None:
```
If this fails on startup you will have to restart HASS. So you might get a race condition if your emoncms server starts after HASS

maybe rather increase seconds if it is None and then it will recover

```python
secs = 600 if self.data is None else self._interval
...snip...
now+timdelta(seconds=secs)
```

joyrider3774 replied:
Not sure how i can make it work then in combination with what setup_platform does as it needs the data in order for it to be able to check if the feedid was specified in the include_feed_id or exclude_feed_id. When both those config values are not specified the component normally creates sensors for all feedids so i can't create any sensors either as it is not known at that time then how many feeds there are and thus how many sensors need to be created.
Not sure how i can get around these problems then

kellerza commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+"""
+Support for monitoring emoncms feeds.
+
+https://emoncms.org
```
Please refer to the HASS docs here (there you can point to emoncms), similar to:

```
"""
Support for monitoring emoncms feeds.

For more details about this component, please refer to the documentation
at https://home-assistant.io/components/sensor.emoncms/
"""
```

kellerza commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+        found = False
+        if self._data.data is not None:
+            for elem in self._data.data:
+                elemid = get_id(self._sensorid, elem["tag"], elem["name"],
```
It would be nice to build a dict in your update method below with get_id as the key. Then you only do it once for multiple sensors, but that's up to you!

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+    vol.Exclusive(CONF_EXCLUDE_FEEDID, 'only_include_or_exclude'):
+        vol.All(cv.ensure_list, [cv.positive_int]),
+    vol.Optional(CONF_INCLUDE_FEEDID_NAMES):
+        vol.All(cv.ensure_list, [cv.string]),
```
Why not make this a dict that maps the feed ids to the names?

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+
+def get_id(sensorid, feedtag, feedname, feedid, feeduserid):
+    """Return unique identifier for feed / sensor."""
+    return "emoncms" + sensorid + "_" + feedtag + "_" + \
```
String formatting is preferred over string concatenation.
```
`emoncms{}_{}_{}_{}_{}`.format(
    sensorid, feedtag, feedname, feedid, feeduserid)
```

MartinHjelmare replied:
Sorry, I've used back-tics around the string above. Those should be regular single quotes or double quotes depending on your preference. Just be consistent over the module.

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+
+    def update(self, now):
+        """Get the latest data."""
+        try:
```
You can remove the outer `try` statement by doing:
```py
try:
***
except
***
else:
***check status code here***
finally:
***
```

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+
+    for elem in data.data:
+        if include_feeds is None and exclude_feeds is None:
+            sensors.append(EmonCmsSensor(hass, data, None, value_template,
```
Break this out and put it at the end of the loop. Just set name to None here. Then you can remove the other similar statements below.

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+
+        elif exclude_feeds is not None:
+            if not int(elem["id"]) in exclude_feeds:
+                sensors.append(EmonCmsSensor(hass, data, None,
```
See above.
```py
if int(elem[`id`]) in exclude_feeds:
    continue
```

MartinHjelmare commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+                                             elem))
+
+        elif include_feeds is not None:
+            if int(elem["id"]) in include_feeds:
```
See above.
```py
if int(elem[`id`]) not in include_feeds:
    continue
name = None
***etc***
```

joyrider3774 committed:
Fix comment so it refers to the documentation

joyrider3774 committed:
use string formatting

joyrider3774 committed:
Remove outer try

joyrider3774 committed:
clean up sensors.append calls
(break them out for loop)

fabaff commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+    include_feeds = config.get(CONF_INCLUDE_FEEDID)
+    include_feeds_names = config.get(CONF_INCLUDE_FEEDID_NAMES)
+    interval = util.convert(config.get(CONF_SCAN_INTERVAL), int,
+                            DEFAULT_SCAN_INTERVAL)
```
The default for the interval can be set by the validation.

fabaff commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+                 unit_of_measurement, sensorid, elem):
+        """Initialize the sensor."""
+        if name is None:
+            self._name = "emoncms" + sensorid + "_feedid_" + elem["id"]
```
String formatting is preferred.

fabaff commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+https://emoncms.org
+"""
+from datetime import timedelta
+import logging
```
Please add a blank line to divide the built-in modules from the third-party modules.

fabaff commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,246 @@
+from datetime import timedelta
+import logging
+import requests
+import voluptuous as vol
```
Please add a blank line to divide the third-party modules from our imports.

joyrider3774 committed:
multiple changes like:
- rename config value "include_feed_id" to "include_only_feed_id"
- rename config value "include_feed_id_names" to "sensor_names"
- renamed config value sensor_names is now a dictionary an can also be used when renamed config value "include_only_feed_id" is not specified
- Set default value for scan_interval using the config validation
- blank lines between default, 3rd party and own imports
- removed homeassistant.util import, it was not needed anymore

joyrider3774 commented:
guys i did everything except for the possible racing condition that may occur if you run a local emoncms and hass on the same pc. If hass starts before emoncms starts the sensors won't be created since setup_platform will fail then because the url was not availible. I have no idea how i can easily fix this though as the sensors itselve need some of the data (like feedid) as well as setup_platform for inclusion / exclusion of heeds and name mapping

joyrider3774 committed:
fix extended voluptuous schema
scan_interval should not be extended to PLATFORM_SCHEMA as it was already in the schema by config_validation helper so i should not add / change it. It was also causing problems reading the value from the config.

balloob commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,224 @@
+                              CONF_URL, req.status_code)
+        finally:
+            if self.data is not None:
+                track_point_in_utc_time(self._hass,
```
Instead of tracking this yourself, let Home Assistant polling trigger updates. 

- `EmonCmsSensor.update()` calls `EmonCmsData.update()`.
- Add `@Throttle(…)` to `EmonCmsData.update()`.

See [Bitcoin sensor](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/bitcoin.py#L178) for an example

balloob commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,224 @@
+
+    def update(self):
+        """Get the latest data and updates the state."""
+        if self._data is None:
```
This can never happen

balloob commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,224 @@
+            return
+
+        found = False
+        if self._data.data is not None:
```
Better do a guard clause:

```python
if self._data.data is None:
    return
```

balloob commented on homeassistant/components/sensor/emoncms.py:
```diff
@@ -0,0 +1,224 @@
+
+        found = False
+        if self._data.data is not None:
+            for elem in self._data.data:
```
Magic time ;-)

You can simplify your code a bit. The main thing here is to just store the `elem` object as an instance variables, saves you tons of instance variables.

```python
elem = next((elem for elem in self._data.data
             if get_id(self._sensorid, elem["tag"], elem["name"],
                       elem["id"], elem["userid"]) == self.identifier), None)

if elem is None:
    return

self._elem = elem

if self._value_template is not None:
    self._state = template.render_with_possible_json_value(
        self._hass, self._value_template, elem['value'],
        STATE_UNKNOWN)
else:
    self._state = round(float(elem['value']), DECIMALS)
```

joyrider3774 committed:
Use Home Assistant polling

joyrider3774 committed:
remove statement that can never happen

joyrider3774 committed:
Guard clause

joyrider3774 committed:
Reduce instance variables

joyrider3774 committed:
Merge branch 'dev' into emoncms

balloob commented:
Bueno! 🐬

devdelay commented:
This is really cool. Would it be possible to also post existing sensor energy/temp data to emoncms.org?

joyrider3774 commented:
you mean like importing existing data into emoncms.org, not related to the component ? they have an input api https://www.emoncms.org/site/api#input you basically start with input's (nodes) and then you create feeds based on the node data. 

The component only creates sensors for feeds. i could have created sensors for input's instead of only feeds as well but i figured in most cases inputs are only raw data and you want the processed data (feeds). just add a log to feed "action" with the input.

not sure if this is what you wanted to know ?

devdelay commented:
I would like to feed emoncms the data from my zwave energy meter sensors and ecobee temperature sensors.

tinglis1 commented:
I have used emoncms in the past to store power usage data from my zwave devices etc. I was using openhab at the time.
I mainly used emoncms due to it exceptional calculation ability and easy of retrieving large quantities of long term data. And can't forget its charting.

A component based on some of the history ones that already exist and this one shouldn't be too hard to put together.

devdelay commented:
seems like a RESTful Sensor would do the trick but not exactly sure how to get the api key in there

joyrider3774 commented:
append &apikey=YOURWRITEAPIKEY (if you want to write) to the url.

i could create a new component just like the dweet.io history one. It actually is a great idea to have a component that could post data to emoncms as well. you'll basically enable any components data like sensors etc that are compatible with hass to send it's data to emoncms. 

@tinglis1 can it be a seperate component like a emoncms history component ? not sure how i can merge both of them but i could start with a component that posts it's data to a specific (configurable) input node to emoncms and perhaps i (or someone else) can try to merge it ?

tinglis1 commented:
I think separate is the best way.

tinglis1 commented:
@devdelay Probably should move the discussion for a emoncms history component to the forum as a feature request.
